src/relay/op/tensor/reduce.cc

```diff
@@ -119,6 +119,56 @@ Array<Integer> GetExcludeAxes(size_t indim,
   return r_axes;
 }
 
+Array<Array<Layout>> SumInferCorrectLayout(const Attrs& attrs, const Array<Layout>& new_in_layouts,
+                                           const Array<Layout>& old_in_layouts,
+                                           const Array<Array<IndexExpr>>& old_in_shapes) {
+  // NOTE: Discard "const" qualifier here.
+  ReduceAttrs* params = const_cast<ReduceAttrs*>(attrs.as<ReduceAttrs>());
+
+  // Get the reduce axes.
+  uint32_t indim = old_in_shapes[0].size();
+  auto r_axes = GetReduceAxes(indim, params->axis, params->exclude);
+
+  Layout ret = Layout::Undef();
+  if (new_in_layouts.defined() && r_axes.size() && params->keepdims) {
+    // Adapt to new layout. The axis has to change. Record original reduce axes. Convert to the
+    // modified layout axes.
+    CHECK_EQ(new_in_layouts.size(), 1);
+    CHECK_EQ(old_in_layouts.size(), 1);
+
+    // 1) Collect the original axes
+    std::unordered_set<char> old_r_dims;
+    for (auto r_axis : r_axes) {
+      old_r_dims.emplace(old_in_layouts[0].name().at(r_axis));
+    }
+
+    // 2) Collect the new axes by walking new_layout.
+    tvm::Array<tvm::Integer> new_r_axis;
+    std::string new_layout_string = "";
+    int axis_index = 0;
+    for (auto layout_axis : new_in_layouts[0]->axes) {
+      char layout_dim = layout_axis->var->name_hint.at(0);
+      if (old_r_dims.count(layout_dim)) {
+        new_r_axis.push_back(tvm::Integer(axis_index));
+      }
+      // Collect only the primal axis.
+      if (LayoutAxis::Get(layout_axis).IsPrimal()) {
+        new_layout_string += layout_dim;
```

Comment:
how about the sub-axis?

Reply:
Sub axis, we dont have to do anything. Basically, this will insert a layout transform from NCHWC16c to NCHW before reduce.